src/org/omegat/core/machinetranslators/ApertiumTranslate.java

```diff
@@ -83,18 +94,23 @@ private String apertiumCode(Language language) {
 
     @Override
     protected String translate(Language sLang, Language tLang, String text) throws Exception {
-        String prev = getFromCache(sLang, tLang, text);
-        if (prev != null) {
-            return prev;
-        }
-
         String trText = text;
 
         String sourceLang = apertiumCode(sLang);
         String targetLang = apertiumCode(tLang);
 
-        String url2 = GT_URL2.replace("#sourceLang#", sourceLang).replace("#targetLang#", targetLang);
-        String url = GT_URL + URLEncoder.encode(trText, "UTF-8") + url2;
+        String server = getServerUrl();
+        String apiKey = getCredential(PROPERTY_APERTIUM_API_KEY);
+
+        if (!useCustomServer()) {
+            server = PROPERTY_APERTIUM_SERVER;
+        }
+        if (!useCustomServer()) {
+            apiKey = PROPERTY_APERTIUM_API_KEY;
+        }
+
+        String url2 = PROPERTY_APERTIUM_URL2.replace("#sourceLang#", sourceLang).replace("#targetLang#", targetLang);
+        String url = server + PROPERTY_APERTIUM_URL + URLEncoder.encode(trText, "UTF-8") + url2 + apiKey;
```

Comment:
Building the URL by concatenation seems dangerous to me. Instead I would suggest a format string, like:

```java
protected static final String APERTIUM_URL_FORMAT = "https://%s/apy/translate?q=%s&markUnknown=no&langpair=%s|%s&key=%s";

...

String url = APERTIUM_URL_FORMAT.format(server, trText, sourceLang, targetLang, apiKey);
```

Reply:
You are absolutely right, something similar was already in place but it is better to fix it now to prevent future issues. Fixed in latest commit.